src/block_gmres.jl

```diff
@@ -0,0 +1,349 @@
+
+  start_time = time_ns()
+  solver = BlockGmresSolver(A, B, memory)
+  elapsed_time = ktimer(start_time)
+  timemax -= elapsed_time
+  block_gmres!(solver, A, B; M, N, ldiv, restart, reorthogonalization, atol, rtol, itmax, timemax, verbose, history)
+  solver.stats.timer += elapsed_time
+  return solver.X, solver.stats
+end
+
+function block_gmres!(solver :: GmresSolver{T,FC,S}, A::AbstractMatrix{FC}, X0::AbstractMatrix{FC}; M=I, N=I,
+                      ldiv::Bool=false, restart::Bool=false, reorthogonalization::Bool=false,
+                      atol::T = √eps(T), rtol::T=√eps(T), itmax::Int=0,
+                      timemax::Float64=Inf, verbose::Int=0, history::Bool=false) where {T <: AbstractFloat, FC <: FloatOrComplex{T}, S <: AbstractMatrix{FC}}
+
+  start_time = time_ns()
+  warm_start!(solver, X0)
+  elapsed_time = ktimer(start_time)
+  timemax -= elapsed_time
+  block_gmres!(solver, A, B; M, N, ldiv, restart, reorthogonalization, atol, rtol, itmax, timemax, verbose, history)
+  solver.stats.timer += elapsed_time
+  return solver
+end
+
+function block_gmres!(solver :: GmresSolver{T,FC,S}, A, B::AbstractMatrix{FC}; M=I, N=I,
+                      ldiv::Bool=false, restart::Bool=false, reorthogonalization::Bool=false,
+                      atol::T = √eps(T), rtol::T=√eps(T), itmax::Int=0,
+                      timemax::Float64=Inf, verbose::Int=0, history::Bool=false) where {T <: AbstractFloat, FC <: FloatOrComplex{T}, S <: AbstractMatrix{FC}}
+
+  # Timer
+  start_time = time_ns()
+  timemax_ns = 1e9 * timemax
+
+  n, m = size(A)
+  s, p = size(B)
+  m == n || error("System must be square")
+  n == s || error("Inconsistent problem size")
+  (verbose > 0) && @printf("BLOCK-GMRES: system of size %d with %d right-hand sides\n", n, p)
+
+  # Check M = Iₙ and N = Iₙ
+  MisI = (M === I)
+  NisI = (N === I)
+
+  # Check type consistency
+  eltype(A) == FC || @warn "eltype(A) ≠ $FC. This could lead to errors or additional allocations in operator-matrix products."
+  ktypeof(B) <: S || error("ktypeof(B) is not a subtype of $S")
+
+  # Set up workspace.
+  allocate_if(!MisI  , solver, :Q , S, n, p)
+  allocate_if(!NisI  , solver, :P , S, n, p)
+  allocate_if(restart, solver, :ΔX, S, n, p)
+  ΔX, X, W, V, Z = solver.ΔX, solver.X, solver.W, solver.V, solver.Z
+  R, stats = solver.R, solver.stats
+  warm_start = solver.warm_start
+  RNorms = stats.residuals
+  reset!(stats)
+  Q  = MisI ? W : solver.Q
+  R₀ = MisI ? W : solver.Q
+  Xr = restart ? ΔX : X
+
+  # FIX ME
+  Ψtmp = Ztmp = S(undef, p, p)
+  τ = S(undef, p)
+
+  # Coefficients for mul!
+  α = -one(FC)
+  β = one(FC)
+  γ = one(FC)
+
+  # Initial solution X₀.
+  X .= zero(FC)
+
+  # Initial residual R₀.
+  if warm_start
+    mul!(W, A, Δx)
+    W .= B .- W
+    restart && (X .+= ΔX)
+  else
+    W .= B
+  end
+  MisI || mulorldiv!(R₀, M, W, ldiv)  # R₀ = M(B - AX₀)
```

Comment:
what is the purpose of the function `mulorldiv!`? I think that name is slightly confusing.

Reply:
It's for the preconditioners. Depending on the option provided by the user, mul! or ldiv! is used.
https://jso.dev/Krylov.jl/dev/preconditioners/